### ISSUE TITLE
Pickling of case class requires more than picklers of its members

### ISSUE THREAD

**Issue by Gralfim** [benhutchison/prickle#37]
In the attached example there is a design pattern we wants to use in our project. Unpickling is OK, pickling of case class Model(list: List[Vector]) works, but pickling of case class Model2(origin: Vector) does not compile. Is there any change to fix it?

build.sbt:
```
name := "JacksonModelTypes"
version := "1.0"
scalaVersion := "2.12.0"
libraryDependencies += "com.github.benhutchison" %% "prickle" % "1.1.13"
```

ModelTypes.scala:
```
import prickle._

object MT {
  val EngineInstance: ModelTypes = jmeModelTypes
}

trait ModelTypes {
  type Vector

  trait ConstructVector {
    def apply(): Vector
    def apply(x: Float, y: Float): Vector
  }
  implicit val Vector: ConstructVector

  def getVectorPickler: Pickler[Vector]
  implicit val vectorPickler: Pickler[Vector] = getVectorPickler
  def getVectorUnpickler: Unpickler[Vector]
  implicit val vectorUnpickler: Unpickler[Vector] = getVectorUnpickler
}

object jmeModelTypes extends ModelTypes {
  type Vector = JME.Vector

  override def getVectorPickler: Pickler[Vector] = Pickler.materializePickler[Vector]
  override def getVectorUnpickler: Unpickler[Vector] = Unpickler.materializeUnpickler[Vector]

  object Vector extends ConstructVector {
    def apply() = JME.Vector(0,0)
    def apply(x: Float, y: Float) = JME.Vector(x,y)
  }
}

object JME {
  case class Vector(x: Float, y: Float)
}

object Main {

  object Abstract {
    import MT.EngineInstance._

    case class Model(list: List[Vector])
    def getJson: String = {
      val model = Model(List(Vector(1, 2), Vector(3, 4), Vector(1.2f, 3.14f)))
      Pickle.intoString(model)
    }

    case class Model2(origin: Vector)
    def getJson2: String = {
      val model = Model2(Vector(1, 2))
      Pickle.intoString(model)
    }
  }
}

```

**Comment by benhutchison** [comment]
Can you include the compiler error, or better yet link me a failing Scastie  http://scastie.org

There no particular reason it shouldn't work. Ambiguous implicits might explain.  Calling materialize pickler manually shouldn't be necessary. Follow the basic  readme format,  see if it works, then try the design pattern

**Comment by Gralfim** [comment]
link: http://scastie.org/25978
errors: 
Error:(52, 24) could not find implicit value for parameter p: prickle.Pickler[Main.Abstract.Model2]
Error occurred in an application involving default arguments.
      Pickle.intoString(model)
Error:(52, 24) not enough arguments for method intoString: (implicit p: prickle.Pickler[Main.Abstract.Model2], implicit config: prickle.PConfig[microjson.JsValue])String.
Unspecified value parameters p, config.
Error occurred in an application involving default arguments.
      Pickle.intoString(model)

**Comment by benhutchison** [comment]
I dont think its ambiguous implicits. Rather its the use of type aliasing
of Vector. The Picklers can only support un-aliased  case classes .

**Comment by OndrejSpanel** [comment]
This is not very clear to me. Pickling and unpickling of `Vector3` alone works fine. It is only pickling case class containing the `Vector3` which breaks thinks.

**Comment by benhutchison** [comment]
Can you post Scasties for what you mean. Theres no Vector3 the in the last
eg posted.

I believe the problem is the type alias on line 29 of last example. Can you
show a failing example that doesnt alias the case class?

**Comment by OndrejSpanel** [comment]
Sorry. `Vector3` was just a typo (edited). (BTW: is it necessary to post the quotes including signatures in your replies, it makes the issue kind of long?)

We understand aliasing is related to the issue and there is no issue without aliasing, however we think we need aliasing to be able to get abstraction over `Vector`, therefore this is something which is not easy to get rid for us on our side.

 Even the aliased `Vector` works fine when picked alone or as a part of the `List`. It does not work only when `Vector` is a direct member of a case class. This is what the Scastie already shows.

**Comment by benhutchison** [comment]
Unpickling an aliased type just isnt possible. What constructor would the code generated by the unpickling macro invoke to instantiate the class? You cannot create an instance of an alias.

**Comment by OndrejSpanel** [comment]
In all due respect, is not the closer a bit hasty? I understand we may describe the issue in a not very clear way, and the issue is a corner case, but still I think the issue is real.

Note:
- the same `Vector` can be serialized fine when model is defined as `case class Model(list: List[Vector])`, but not when it is `case class Model(origin: Vector)`
- Jackson serializes the same structures fine with no issues

I think this both indicates the constructor to use is well defined and should be discoverable.

**Comment by benhutchison** [comment]
If you think you can find a solution then please send a PR. Here's the section you'd need to redesign, if you have any solution proposals happy to assist integrating them.

https://github.com/benhutchison/prickle/blob/master/shared/src/main/scala/prickle/PicklerMaterializers.scala#L14-L17

**Comment by benhutchison** [comment]
Fixed by #38, released as `1.1.14`
